### ISSUE TITLE
[expo] Dragging a window from a workspace to another makes the window disappear

### ISSUE THREAD

**Issue by goingtosleep** [WayfireWM/wayfire#1869]
**Describe the bug**
When I drag a window (suspect it only happens with xwayland, I haven't encountered this with native wayland windows) from a workspace to another, sometimes it will shrink the window and disappear (see the video).

Also looks like only dragging a maximized window has this issue. I tried with a non-maximized vivaldi and it's fine.
~Update: Seems fine with client-side decoration.~ Happens with client-side decoration too.

**To Reproduce**
1. Enable expo
2. Open an xwayland window, in the video it's `vivaldi-stable`.
3. Activate expo, drag the window from a workspace to another.
4. The window will disappear.

**Expected behavior**
The window will be moved to the target workspace without issues.

**Screenshots or stacktrace**

https://github.com/WayfireWM/wayfire/assets/47395035/ae3175ef-83c7-47c2-ac96-5efd4c8fb6cb

**Wayfire version**
wayfire-git

**Comment by ammen99** [comment]
Could you collect debug logs from Wayfire for this issue with `-d txn -d txni -d xwayland -d views`? I.e. start wayfire with the given additional flags (and redirect stdout and stderr to a file) and then reproduce the issue, and then share the log file. Preferably stop wayfire immediately after the bug happens (ctrl-alt-backspace) so that we don't get a ton of unnecessary output afterwards.

**Comment by goingtosleep** [comment]
Not sure why it's not as easy to reproduce as when I issued the bug report, but still got the issue twice as below, please have a look @ammen99:

[debug.txt](https://github.com/WayfireWM/wayfire/files/12447983/debug.txt)
[debug2.txt](https://github.com/WayfireWM/wayfire/files/12447984/debug2.txt)

**Comment by ammen99** [comment]
I think I figured out what the problem is, it is actually an issue in Vivaldi and has nothing to do with Expo. The easiest way to reproduce is this: open vivaldi, maximize and close it. Start again (so that it is maximized), try to unmaximize with the grid plugin. What happens:

Wayfire tries to restore the window to its 'native' size. For Xwayland, this means resizing a window to its base size as indicated in the hints (https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/wm-normal-hints.html). The base size is equal to min size if not otherwise specified. And Vivaldi says its min size is 1x1. Therefore, Wayfire assumes that the default size of the browser is 1x1 and resizes it to that size. IMO this is a Vivaldi bug. I suppose a plugin could add a workaround for this (e.g. overwrite the base size of the vivaldi browser), but I don't think such a thing belongs in Wayfire upstream.

TLDR: this is a bug in Vivaldi, their minimal size (1x1) does not really make much sense as a default size.

**Comment by ammen99** [comment]
As an easy workaround, you can run the following small program periodically (e.g. every 1 second or so):

```cpp
// Compile with g++ x11.cpp -o x11 `pkg-config --cflags --libs x11`
#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <cmath>
#include <algorithm>
#include <cstring>

int main() {
    auto display = XOpenDisplay(nullptr);
    auto root = XDefaultRootWindow(display);
#define MINSIZE 300

    Window tmp1, tmp2;
    Window *children;
    unsigned int n_children;
    XQueryTree(display, root, &tmp1, &tmp2, &children, &n_children);
    for (unsigned int i = 0; i < n_children; i++) {
        XSizeHints hints;
        long flags;
        XGetWMNormalHints(display, children[i], &hints, &flags);

        bool update = false;

        if (hints.min_width < MINSIZE || hints.min_height < MINSIZE) {
            hints.min_width = std::max(hints.min_width, MINSIZE);
            hints.min_height = std::max(hints.min_height, MINSIZE);
            hints.flags |= PMinSize;
            update = true;
        }

        if (hints.base_width < MINSIZE || hints.base_height < MINSIZE) {
            hints.base_width = std::max(hints.base_width, MINSIZE);
            hints.base_height = std::max(hints.base_height, MINSIZE);
            hints.flags |= PBaseSize;
            update = true;
        }

        if (update) {
            XSetWMNormalHints(display, children[i], &hints);
            XFlush(display);
        }
    }

    XCloseDisplay(display);
    return 0;
}
```

It will go through all toplevel Xwayland windows and set their base/min sizes to 300x300 (adjust if needed), so that they are not resized to a small default size.
